src/gromacs/nblib/topology.cpp

```diff
@@ -188,13 +189,22 @@ std::vector<real> TopologyBuilder::extractAtomTypeQuantity(Extractor extractor)
 Topology TopologyBuilder::buildTopology()
 {
     topology_.excls_  = createExclusionsList();
-    topology_.masses_ = extractAtomTypeQuantity(
+    topology_.masses_ = extractAtomTypeQuantity<real>(
             [](const auto& data, auto& map) { return map[data.atomTypeName_].mass(); });
-    topology_.charges_ = extractAtomTypeQuantity([](const auto& data, auto& map) {
+    topology_.charges_ = extractAtomTypeQuantity<real>([](const auto& data, auto& map) {
         ignore_unused(map);
         return data.charge_;
     });
 
+    topology_.atomTypes_ = extractAtomTypeQuantity<std::string>(
+            [](const auto& data, auto& map) { return map[data.atomTypeName_].name(); });
+    //! Create a temporary list to sort the AtomType names and remove duplicates
+    std::list<std::string> atomTypesList(std::begin(topology_.atomTypes_), std::end(topology_.atomTypes_));
+    atomTypesList.sort();
+    atomTypesList.unique();
+    topology_.uniqueAtomTypes_.insert(std::end(topology_.uniqueAtomTypes_),
+                                      std::begin(atomTypesList), std::end(atomTypesList));
+
```

Comment:
The list here is not really the best tool to achieve uniqueness and sortedness. You can achieve that
much more directly and efficiently by operating on a vector.
```suggestion
auto& uvec = topology_.uniqueAtomTypes_;
uvec = topology_.atomTypes_;
std::sort(std::begin(uvec), std::end(uvec));
auto unique_end = std::unique(std::begin(uvec), std::end(uvec));
uvec.erase(unique_end, std::end(uvec));
```

Reply:
We might want to move the above code into an independent utility function - this could also be used inside
molecule.

Reply:
Seconded. Could be a free utility function. Even if the utility is used only in this class, I'd not turn it into a private member function.